### PR TITLE
filter commits in git.Commits based on sub-repo permissions

### PR DESCRIPTION
**Pull request by mollylogue**
This PR adds sub-repo permissions filtering on the `git.Commit` function by passing the `--name-only` flag when running `git log` to get the list of files modified in each commit along with the rest of the commit information. I decided to put this part up for review before implementing/testing the filtering for the other functions in this package so that I could get a signoff on this approach (or feedback on how to do it better) before moving forward. 

https://github.com/sourcegraph/sourcegraph/issues/27498

### CODE REVIEW

File: internal/vcs/git/commits_test.go [hunk 1]
```diff
@@ -314,26 +316,80 @@ func TestRepository_Commits(t *testing.T) {
...
+		"git add file2.2",
+		"GIT_COMMITTER_NAME=c GIT_COMMITTER_EMAIL=c@c.com GIT_COMMITTER_DATE=2006-01-02T15:04:07Z git commit -m commit2 --author='a <a@a.com>' --date 2006-01-02T15:04:06Z",
+		"touch file3",
+		"git add file3",
+		"GIT_COMMITTER_NAME=c GIT_COMMITTER_EMAIL=c@c.com GIT_COMMITTER_DATE=2006-01-02T15:04:07Z git commit -m commit3 --author='a <a@a.com>' --date 2006-01-02T15:04:07Z",
+	}
+
+	tests := map[string]struct {
+		repo        api.RepoName
+		opt         CommitsOptions
```
Comment (ryanslade) [inline internal/vcs/git/commits_test.go:360]: Don't seem to use this

File: internal/vcs/git/commits.go [hunk 2]
```diff
@@ -295,6 +354,9 @@ func commitLogArgs(initialArgs []string, opt CommitsOptions) (args []string, err
 	if opt.Path != "" {
 		args = append(args, "--", opt.Path)
 	}
+	if opt.NameOnly {
```
Comment (ryanslade) [inline internal/vcs/git/commits.go:357]: Can you double check that it works if they've added a non-empty path? I'm not sure how a `--` with no name parameter will play with more parameters following it.

**Review by ryanslade — APPROVED** [review]
LGTM

I wonder how bad the performance would be to always return file names? It may simplify the code quite a bit

**Comment by mollylogue** [comment]
> LGTM
> 
> I wonder how bad the performance would be to always return file names? It may simplify the code quite a bit

I was wondering about that too, but I wanted to make the change as low-impact as possible. I'm open to making it the default, though.

File: internal/vcs/git/commits_test.go (hunk 1 above)
Comment (mollylogue) [reply internal/vcs/git/commits_test.go]: Ah yeah that was for potential future testing.

File: internal/vcs/git/commits.go (hunk 2 above)
Comment (mollylogue) [reply internal/vcs/git/commits.go:353]: So, it does work, but I needed to specify the path last and the `--name-only` flag with the rest of the flags beforehand. I've added tests to demonstrate this behavior.

**Comment by mollylogue** [comment]
@ryanslade I believe I've addressed your comments thus far. I've added to the test cases to have better coverage and validation (specifically around the case you mentioned w/ both `--read-only` and a path specified). One outstanding question is whether or not the default client has been initialized already for all of these use cases. I'm not entirely sure about that.

**Comment by ryanslade** [comment]
> @ryanslade I believe I've addressed your comments thus far. I've added to the test cases to have better coverage and validation (specifically around the case you mentioned w/ both `--read-only` and a path specified). One outstanding question is whether or not the default client has been initialized already for all of these use cases. I'm not entirely sure about that.

Great, LGTM.

I'm not too worried about whether we've initialised everything yet as that will come out during testing and it doesn't matter yet as it's all disabled in prod anyway.

**Comment by ryanslade** [comment]
> > LGTM
> > I wonder how bad the performance would be to always return file names? It may simplify the code quite a bit
> 
> I was wondering about that too, but I wanted to make the change as low-impact as possible. I'm open to making it the default, though.

I just tested locally running `git log` against our sourcegraph repo and adding `--name-only` is significantly slower so I think your approach is worth it.

```
Benchmark 1: git log --name-only > blah.txt
  Time (mean ± σ):      2.102 s ±  0.057 s    [User: 1.929 s, System: 0.156 s]
  Range (min … max):    2.030 s …  2.182 s    10 runs

Benchmark 1: git log > blah.txt
  Time (mean ± σ):     277.9 ms ±   5.0 ms    [User: 249.2 ms, System: 25.3 ms]
  Range (min … max):   269.3 ms … 286.2 ms    10 runs
```

**Review by ryanslade — APPROVED** [review]

**Review by sashaostrikov — APPROVED** [review]
LGTM!

**Comment by mollylogue** [comment]
> > > LGTM
> > > I wonder how bad the performance would be to always return file names? It may simplify the code quite a bit
> > 
> > 
> > I was wondering about that too, but I wanted to make the change as low-impact as possible. I'm open to making it the default, though.
> 
> I just tested locally running `git log` against our sourcegraph repo and adding `--name-only` is significantly slower so I think your approach is worth it.
> 
> ```
> Benchmark 1: git log --name-only > blah.txt
>   Time (mean ± σ):      2.102 s ±  0.057 s    [User: 1.929 s, System: 0.156 s]
>   Range (min … max):    2.030 s …  2.182 s    10 runs
> 
> Benchmark 1: git log > blah.txt
>   Time (mean ± σ):     277.9 ms ±   5.0 ms    [User: 249.2 ms, System: 25.3 ms]
>   Range (min … max):   269.3 ms … 286.2 ms    10 runs
> ```

Oof. ☹️  I really wish that I could've figured out a way to do the filtering directly in the path field (i.e. https://stackoverflow.com/questions/5685007/making-git-log-ignore-changes-for-certain-paths) but it didn't seem like the format was compatible with the path excludes values in the sub-repo permissions (and sub-repo permissions can have multiple path excludes rules). That would've probably been the most performant way to do this, but oh well 🤷‍♀️

**Pull request merged**